src/coreclr/jit/gentree.cpp

```diff
@@ -3779,14 +3779,11 @@ bool GenTreeOp::IsValidLongMul()
     GenTree* op1 = gtGetOp1();
     GenTree* op2 = gtGetOp2();
 
-    if (!TypeIs(TYP_LONG))
+    if (!TypeIs(TYP_LONG) || !op1->TypeIs(TYP_LONG) || op2->TypeIs(TYP_LONG))
     {
         return false;
     }
 
-    assert(op1->TypeIs(TYP_LONG));
-    assert(op2->TypeIs(TYP_LONG));
-
```

Comment:
What prompted this change? Did you see `TYP_BYREF` `op1` or `op2` on arm64? How do we end up with those?

Reply:
Also seems like `op2->TypeIs(TYP_LONG)` should be `!op2->TypeIs(TYP_LONG)`?

Reply:
> What prompted this change? Did you see `TYP_BYREF` `op1` or `op2` on arm64? How do we end up with those?

Both operands are of `TYP_INT`, while the final MUL result should be `TYP_LONG`. The requirement of operands being `TYP_INT` , because (the offset) when they gets patched by relocation, they need to be read as `TYP_INT`. Initially I had it to be `TYP_IMPL`, but that was reading the wrong offset. Likewise, I cannot change the result of `TYP_LONG` to `TYP_INT`, because the requirement is that the result should be 64-bits.

```
N006 ( 10, 15) [000026] -------N---                            |     \--*  MUL       long  
N004 (  5, 10) [000024] #----------                            |        +--*  IND       int   
N003 (  2,  8) [000023] H----------                            |        |  \--*  CNS_INT(h) ref   
N005 (  1,  2) [000025] -----------                            |        \--*  CNS_INT   int    8
```



> Also seems like `op2->TypeIs(TYP_LONG)` should be `!op2->TypeIs(TYP_LONG)`?

Ah, that's why seeing those failures. Will fix it.

Reply:
Where is a `MUL` like this one created? And what are the semantics of this operation supposed to be? It looks like illegal IR to me, e.g. we have TODO in `fgDebugCheckTypes` about checking for mismatched types like in your example.

Can the code that is creating this IR just create `GT_MUL_LONG` which already represents int*int -> long multiplication? Or can it insert the appropriate casts on the operands to allow the transformation here to pick it up? I do not think we should allow the operands of `GT_MUL` to be mismatched with the result type.

Reply:
Changed it to `LSH(x, 3)` works better.